### ISSUE TITLE
Network compartment not created until a container is started in the HCN namespace

### ISSUE THREAD

**Issue by rawahars** [microsoft/Windows-Containers#401]
**Describe the bug**
We are following the standard workflow of running containers using `containerd` runtime wherein we-
- Create an HCN namespace
- Configure the same using CNI plugin
- Start a container using containerd and pass the HCN namespace ID in the OCI spec options.

We are observing after step 2 that even though the HCN namespace seems to be configured (by checking the output of `hnsdiag`), the actual network compartment is not created (checked by running `Get-NetCompartment`).
Once the container has been started in the same HCN namespace then the network compartment is created and can be viewed via `Get-NetCompartment`.

**To Reproduce** 
- Create an HCN namespace
- Configure the same using CNI plugin (Hack the CNI plugin to sleep for 15 seconds once it is finished).
- Check the HNS resources using `hnsdiag list all` and the network compartment using `Get-NetCompartment`
- Start a container using containerd and pass the HCN namespace ID in the OCI spec options.
- Check the HNS resources using `hnsdiag list all` and the network compartment using `Get-NetCompartment`

**Expected behavior**
I am expecting that the network compartment is created as soon as the HCN namespace is created.

**Configuration:**
 - Edition: Windows Server 2019
- Base Image being used: Windows Server
 - Container engine: containerd
 - Container Engine version 1.6

**Comment by MikeZappa87** [comment]
Hello, is this in the context of Kubernetes (are you using the cri plugin of containerd) or outside of Kubernetes?

**Comment by rawahars** [comment]
Hi @MikeZappa87, No this is outside of Kubernetes environment. 

In Kubernetes, I think pause container is started first in the HCN namespace and then the containers from the pod are launched.

**Comment by MikeZappa87** [comment]
Is it possible to make use of the cri-api (RunPodSandbox/CreateContainer)? This is common outside of the K8s space so you wouldn't be outside of the normal. It would satisfy creating the compartment, executing the cni plugin. I will check why creating a namespace doesn't show in the get-netcompartment.

How are you doing this? Via ctr?

**Comment by rawahars** [comment]
So unfortunately we have our own workflow which does not use the cri-api. Presently I was using the containerd SDK client to accomplish the workflow I mentioned earlier.

However, there is an easy test for this is just using `New-HnsNamespace` Powershell command to create an HCN namespace. If you check the compartments using `Get-NetCompartment`, it won't have any entry. Or if you use `Get-HNSNamespace`, then there won't be a CompartmentID associated with this HCN Namespace.

**Comment by MikeZappa87** [comment]
One item to note is: 

"Start a container using containerd and pass the HCN namespace ID in the OCI spec options." Is that it probably won't do anything since https://github.com/containerd/containerd/blob/4d2c8879908285454a4006534cb0af82bb58a406/pkg/netns/netns_windows.go#L33

We don't pass anything to the create compartment.

**Comment by kestratt** [comment]
HNS will create the compartment during container creation. It is expected that the compartment isn't present yet.

**Comment by kestratt** [comment]
There is work currently being done to tie the lifetime of the network namespace and compartment to the pod itself. This feature is targeting Windows Server 2022.

**Comment by rawahars** [comment]
Thanks @MikeZappa87 and @kestratt for the information.

**Comment by keithmattix** [comment]
@kestratt @fady-azmy-msft is there an ETA for when this feature will be launched? I'm not seeing it in WS2022
